### PR TITLE
Allow SRV as scheme

### PR DESCRIPTION
**Pull request by allevo**
Implement SRV resolution.

This PR implements the SRV resolution allowing the developer to specify a connection url `mongodb+srv://<>`. This kind of url is the Atlas suggested one.
The username and password are no supported in the url because the auth mechanism SCRAM are not supported yet.

Anyway I have write an example for that.

### CODE REVIEW

**Review by jcdyer — COMMENTED** [review]
This looks good, and will be very useful for me personally in about six months (when my project migrates to Atlas). Some suggestions for cleaning up the code below.  

In addition to the line comments, Consider, instead of passing around `is_srv`, and complicating the logic of the `parse` function at every step, split once at the beginning, and handle each connection type in its own `parse_{connection_type}` function.  That way, the logic for each code path follows clearly the specification at https://docs.mongodb.com/manual/reference/connection-string/.

Using the terminology from that page, the top-level parse function might look like:

```
pub fn parse(address: &str) -> Result<ConnectionString> {
    if address.starts_with(URI_SCHEME) {
        parse_standard_connection_str(address)
    } else if address.starts_with(URI_SRV_SCHEME) {
        parse_dns_seedlist_connection_str(address)
    }
}
```
I don't have commit rights here, so I'll have to defer to @saghm or @kyeah for final approval, but I'm happy to re-review with the suggested changes.

**Comment by jcdyer** [comment]
Note: I haven't verified the correctness of the srv resolution, just reviewed the code for structural concerns.

File: src/connstring.rs [hunk 1]
```diff
@@ -333,3 +361,35 @@ fn rsplit<'a>(string: &'a str, sep: &str) -> (&'a str, &'a str) {
...
+        .map(|entity| {
+            let port: u16 = entity.port();
+            let hostname = entity.target();
+            let mut hostname = hostname.to_utf8();
+            hostname.pop();
+            Host::new(hostname, port)
+        })
+        .collect();
+
+    hosts.sort_by(|a, b| a.host_name.partial_cmp(&b.host_name).unwrap());
```
[the remark this answers is not in the document]
Comment (allevo) [reply src/connstring.rs]: No this can be moved in the test where the order is useful.

**Comment by allevo** [comment]
@jcdyer there is a test for the correctness. That test connects to mongo Atlas (without username and password).

https://github.com/allevo/mongo-rust-driver-prototype/blob/f8723ede45ab41b4038de4de1472ecb9fa32fd02/tests/client/connstring.rs#L258-L275

And an example

File: src/lib.rs [hunk 2]
```diff
@@ -148,6 +148,7 @@ extern crate sha1;
 extern crate hmac;
 extern crate pbkdf2;
 extern crate hex;
+extern crate trust_dns_resolver;
```
[the remark this answers is not in the document]
Comment (allevo) [reply src/lib.rs]: I used the same convention. If you want I can remove this

File: src/connstring.rs [hunk 3]
```diff
@@ -100,14 +103,20 @@ impl ConnectionString {
...
 
+    let is_srv = address.starts_with(URI_SRV_SCHEME);
+
     // Remove scheme
-    let addr = &address[URI_SCHEME.len()..];
+    let addr = if !is_srv {
+        &address[URI_SCHEME.len()..]
+    } else {
+        &address[URI_SRV_SCHEME.len()..]
+    };
```
[the remark this answers is not in the document]
Comment (allevo) [reply src/connstring.rs]: good spot!

File: src/connstring.rs [hunk 4]
```diff
@@ -163,8 +180,14 @@ pub fn parse(address: &str) -> Result<ConnectionString> {
...
     // Collect options if any exist
-    if !opts.is_empty() {
-        options = Some(split_options(opts).unwrap());
+    if !opts.is_empty() || is_srv {
+        options = Some(split_options(opts, is_srv).unwrap());
+    }
+
+    if is_srv && (user.is_some() || password.is_some()) {
+        return Err(ArgumentError(String::from(
+            "username and password are not currently supported in url.",
```
[the remark this answers is not in the document]
Comment (allevo) [reply src/connstring.rs]: What do you think to suggest the workaround in README?
Maybe linking the example

File: src/connstring.rs [hunk 5]
```diff
@@ -142,9 +151,17 @@ pub fn parse(address: &str) -> Result<ConnectionString> {
...
+        } else {
+            split_hosts(host_string)?
+        };
     } else {
-        hosts = split_hosts(host_str)?;
+        hosts = if is_srv {
+            resolve_srv(host_str)?
+        } else {
+            split_hosts(host_str)?
+        };
```
[the remark this answers is not in the document]
Comment (allevo) [reply src/connstring.rs]: I cannot assign the "return value" of an `if`. Anyway, I have rework this piece

File: src/connstring.rs [hunk 6]
```diff
@@ -288,7 +311,7 @@ fn parse_options(opts: &str, delim: Option<&str>) -> ConnectionOptions {
 }
 
 // Determines the option delimiter and offloads parsing to parse_options.
-fn split_options(opts: &str) -> Result<ConnectionOptions> {
+fn split_options(opts: &str, is_srv: bool) -> Result<ConnectionOptions> {
```
[the remark this answers is not in the document]
Comment (allevo) [reply src/connstring.rs]: addressed partially.

[the remark this answers is not in the document]
Comment (allevo) [reply src/connstring.rs]: Addressed

File: src/connstring.rs [hunk 7]
```diff
@@ -333,3 +361,35 @@ fn rsplit<'a>(string: &'a str, sep: &str) -> (&'a str, &'a str) {
...
+fn map_thrust_dns_error(error: trust_dns_resolver::error::ResolveError) -> Error {
+    Error::DefaultError(error.to_string())
+}
+
+// Resolve SRV entry.
+fn resolve_srv<'a>(host_string: &'a str) -> Result<Vec<Host>> {
+    let resolver = Resolver::default().map_err(map_io_error)?;
+
+    let dns_entry = "_mongodb._tcp.".to_owned() + host_string;
+    let srv_lookup = resolver.lookup_srv(&dns_entry).map_err(map_thrust_dns_error)?;
```
[the remark this answers is not in the document]
Comment (allevo) [reply src/connstring.rs]: I have removed and an error is shown as below:
```
error[E0277]: `?` couldn't convert the error to `error::Error`
   --> src/connstring.rs:399:53
    |
399 |     let srv_lookup = resolver.lookup_srv(&dns_entry)?;
    |                                                     ^ the trait `std::convert::From<trust_dns_resolver::error::ResolveError>` is not implemented for `error::Error`
    |
    = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
    = help: the following implementations were found:
              <error::Error as std::convert::From<&'a str>>
              <error::Error as std::convert::From<bson::DecoderError>>
              <error::Error as std::convert::From<bson::EncoderError>>
              <error::Error as std::convert::From<bson::oid::Error>>
            and 6 others
    = note: required by `std::convert::From::from`

error: aborting due to previous error

```

File: src/connstring.rs [hunk 8]
```diff
@@ -333,3 +361,35 @@ fn rsplit<'a>(string: &'a str, sep: &str) -> (&'a str, &'a str) {
...
+}
+
+// Map trust_dns_resolver into Error
+fn map_thrust_dns_error(error: trust_dns_resolver::error::ResolveError) -> Error {
+    Error::DefaultError(error.to_string())
+}
+
+// Resolve SRV entry.
+fn resolve_srv<'a>(host_string: &'a str) -> Result<Vec<Host>> {
+    let resolver = Resolver::default().map_err(map_io_error)?;
```
[the remark this answers is not in the document]
Comment (allevo) [reply src/connstring.rs]: Done

File: src/connstring.rs [hunk 9]
```diff
@@ -333,3 +361,35 @@ fn rsplit<'a>(string: &'a str, sep: &str) -> (&'a str, &'a str) {
...
+
+// Map io into Error
+fn map_io_error(error: std::io::Error) -> Error {
+    Error::DefaultError(error.to_string())
+}
+
+// Map trust_dns_resolver into Error
+fn map_thrust_dns_error(error: trust_dns_resolver::error::ResolveError) -> Error {
+    Error::DefaultError(error.to_string())
+}
```
[the remark this answers is not in the document]
Comment (allevo) [reply src/connstring.rs]: partially addresses. See below

File: src/connstring.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jcdyer) [reply src/connstring.rs]: That sounds reasonable.

File: src/connstring.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (allevo) [reply src/connstring.rs]: done

File: src/connstring.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (allevo) [reply src/connstring.rs]: Add the link to README.

**Comment by saghm** [comment]
Hello everyone! We've just released an alpha of [an official MongoDB Rust driver](https://github.com/mongodb/mongo-rust-driver); you can read about it [here](https://www.mongodb.com/blog/post/announcing-the-official-mongodb-rust-driver)! Because of this, we won't be updating this driver anymore, and the repository will be made read-only as soon as I get finished posting these messages. Thanks for using this driver, and we hope you try out the new one!

**Pull request closed without merging**